### ISSUE TITLE
Limited Dictionary

### ISSUE THREAD

**Issue by szalkas** [eldoogy/PowerData#6]
I started to play with your amazing code, unfortunately iOS 10 messed up things. However I was able to get the limited dictionary through dynamically linking IOKit library which is fine but not App Store friendly. Is it somehow possible to catch this limited dictionary using the original approach, without messing with IOKit directly? I spent several hours debugging and trying to modify your code but it simply doesn't catch anything and the Allocations are empty after the batteryState/batteryLevel fired...

**Comment by eldade** [comment]
I haven't looked at this in a while, but on the iOS 10 betas, my conclusion was that it was unlikely we could get this to work. Honestly the changes in `UIDevice` almost seem as though they are meant to specifically block this code.

The problem is simple: Before, they were loading the dictionary from IOKit, loading the relevant values (battery level, etc.), and setting the `UIDevice` properties (which was invoking our KV observer). Then, inside the KV observer we were locating the dictionary allocation and stealing it.

In iOS 10, by the time the KV observer is called, the dictionary has already been released, preventing us from grabbing it. I have not found an alternate approach for getting to the data without directly invoking IOKit.

The only alternative I could offer would be to obfuscate your dynamic linking into IOKit. A while ago I created a class that effectively obfuscates dynamic linking in Objective-C. Anyone could still see the IOKit calls if they debugged your program, but your calls would be invisible in any kind of static analysis. Please let me know if there is any interest in that library and I will upload it to GitHub.

**Comment by szalkas** [comment]
This sounds interesting, I've made a dynamic linked version of this using IOKit but didn't believe it would fit into the AppStore, but according to what you wrote and what I experienced with iOS10 it seems there is no other way and also it seems the currently exist batter monitoring solutions using the same method for a while now without getting banned. Of course this could change in any minute but still, interesting. I'm interested in your dynamic linking code however, I'm pretty sure I can learn from it a lot. Thanks in advance if you would share it.

**Comment by sergey-zhuravel** [comment]
Guys that something has changed in that time? @eldade I also interesting to look at your decision ;)
as well as the Battery Life app get data on temperature, charger power, and amperage? Thx

**Comment by mice007** [comment]
@zhuravel-sergey thats a nice find... I've just tried Battery Life and yes, it displays temperature data too which is not exist in the limited dict as well... very interesting. 
Update: it also displays the adapter info.

**Comment by eldade** [comment]
I'm looking at Battery Life, and it doesn't seem to show neither temperature nor anything to do with charging. We're talking about this one? https://itunes.apple.com/us/app/battery-life-your-battery/id1166884133?mt=8

**Comment by sergey-zhuravel** [comment]
@eldade Yes! Menu/Battery Details/ section Battery Data and section Charger Information
![img_0868](https://cloud.githubusercontent.com/assets/7910769/22042075/43c4762c-dd12-11e6-921c-d7e5f48a699c.PNG)

**Comment by eldade** [comment]
@zhuravel-sergey That's weird, all I see is charging state (which is available and documented in `UIDevice`). Not seeing anything about temperature (I tried on an iPhone 7 Plus with iOS 10.2)

I wonder if they might be faking it? Try putting your device in the fridge for 10 minutes or so and see if it drops...

**Comment by mice007** [comment]
@eldade yeah this is what I thought too, just finished the outdoor testing in 0°C and after about 10 minutes it still showing 30 - 32°C... so it is fake.

**Comment by eldade** [comment]
@mice007 Yep, fake... I ran several such tests with the real Temperature key (pre iOS 10), and 10 minutes in the fridge caused a huge drop in battery temperature (don't have anywhere near 0°C outdoors where I'm at...)

**Comment by sergey-zhuravel** [comment]
@eldade how get the charger information? in Battery Life app is true information

**Comment by eldade** [comment]
@zhuravel-sergey Looking at the screenshot, that looks like a totally different app. Please send the App Store URL. Also, you're definitely testing this on iOS 10 right?

**Comment by sergey-zhuravel** [comment]
I install this app - Battery Life: Your Battery Doctor by Nga Nguyen https://appsto.re/ua/L0tJfb.i

Testing on iPhone 7 plus, iOS 10.2

**Comment by eldade** [comment]
@zhuravel-sergey Got it. The temperature reading is definitely fake (just try cooling your phone, the numbers seem to be random and unrelated to actual temperature), but the charger information is in fact accurate. 

Now that we have an iOS 10 jailbreak it shouldn't be too hard to attach this to lldb and see what they're doing. Anyone feels like doing that?

**Comment by sergey-zhuravel** [comment]
@eldade  No, I do not want to jailbreak my iphone. If you are not difficult, and when it is free time, you can add this information to the EEIOKitListener? Thank you very much! ;)

**Comment by mice007** [comment]
@eldade yeah it should answer the question...  I'll play with it but currently I don't have a jailbroken iOS10 device but will arrange one soon and try lldb.. however I don't have big experience on debugging iOS applications.

**Comment by vincentpeng520** [comment]
The Apple reply me  like this    "Specifically, the github project https://github.com/eldade/UIDeviceListener abuses private dictionaries and is not compliant for apps submitted to the App Store. It would be appropriate to fully remove all functionality seen in this project from this app before resbumtitting for review."

**Comment by mice007** [comment]
@vincentpeng520 did Apple reject your App because of this?

**Comment by vincentpeng520** [comment]
@mice007  YES, I use the "EEIOKitListener" which from the another project ,but it was rejected by apple ,i use this project to commit my app which can't display the data in ios 10,but the result is like this .when i search related app ,like battery life ,it can get the maxcapacity ,voltage ,i think they use iokit framework ,but why didn't reject by apple .i just want to get this data ,beacuse my app just need this

**Comment by catmans1** [comment]
Hi @vincentpeng520 do you find solution for this problem?
